plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+					     buffer[self->vid_offset+1],
+					     buffer[self->vid_offset]);
+	g_debug ("writing Thunderbolt vendor-id %s", dynamic_vendor_id);
+
+	/* get image product-id */
+	dynamic_product_id = g_strdup_printf ("0x%02x%02x",
+					       buffer[self->pid_offset+1],
+					       buffer[self->pid_offset]);
+	g_debug ("writing Thunderbolt product-id %s", dynamic_product_id);
+
+	/* check header size */
+	memcpy (&header_offset, buffer, sizeof (guint32));
+	g_debug ("Header size: %" G_GUINT32_FORMAT, header_offset);
+	if (header_offset > image_size) {
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			     "Image header is too big: %" G_GUINT32_FORMAT,
+			     header_offset);
+		return FALSE;
+	}
+	buffer += header_offset;
+	image_size -= header_offset;
+
+	/* firmware install */
+	if (!fu_dell_dock_usb4_hub_nvm_write(device, buffer, image_size, 0,
+					     error))
+		return FALSE;
+
+	/* dock will reboot to re-read; this is to appease the daemon */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_PAIR);
+	fu_device_set_version (device, dynamic_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_open (FuDevice *device, GError **error)
+{
+	/* FuUsbDevice->open */
+	if (!FU_DEVICE_CLASS (fu_dell_dock_usb4_parent_class)->open (device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_setup (FuDevice *device, GError **error)
+{
+	const gchar *version;
+
+	/* set version from USB NVM */
+	version = fu_dell_dock_usb4_show_nvm_info(device, error);
+	if (version != NULL){
+		fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_PAIR);
+		fu_device_set_version (device, version);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_probe (FuDevice *device, GError **error)
+{
+	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4 (device);
+	g_autofree gchar *devid = NULL;
+
+	devid = g_strdup_printf ("USB\\VID_%04X&PID_%04X&hub",
+			(guint) fu_usb_device_get_vid (FU_USB_DEVICE (device)),
+			(guint) fu_usb_device_get_pid (FU_USB_DEVICE (device)));
+
+	fu_device_set_logical_id (device, "hub");
+	fu_device_add_instance_id (device, devid);
+
+	self->intf_nr = GR_USB_INTERFACE_NUMBER;
+	self->blocksz = GR_USB_BLOCK_SIZE;
+	fu_device_add_protocol (device, "com.dell.dock");
+
+	/* this is true only when connected to non-thunderbolt port */
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_close (FuDevice *device, GError **error)
+{
+	/* FuUsbDevice->close */
+	return FU_DEVICE_CLASS (fu_dell_dock_usb4_parent_class)->close (device, error);
+}
+
+static void
+fu_dell_dock_usb4_finalize (GObject *object)
+{
+	G_OBJECT_CLASS (fu_dell_dock_usb4_parent_class)->finalize (object);
+}
+
+static void
+fu_dell_dock_usb4_init (FuDellDockUsb4 *self)
+{
+	fu_device_add_protocol (FU_DEVICE (self), "com.dell.dock");
```

Comment:
Shouldn't this be `com.intel.thunderbolt`?

Reply:
I don't understand, why we need `com.intel.thunderbolt` plugin in `fu-dell-dock-usb-usb4.c`?

Reply:
Because it's the protocol that is being used to flash the device and to understand the FW image.
It's a Thunderbolt/USB4 device.  See the I2C version uses this too: https://github.com/fwupd/fwupd/blob/master/plugins/dell-dock/fu-dell-dock-i2c-tbt.c#L279

Reply:
My understanding if `dell-dock` plugin adds `thunderbolt` protocol furthermore `thunderbolt` has `priority` set over `dell-dock` then `vfuncs` will be overridden by `thunderbolt` plugin?

Reply:
Well actually you never get into a situation that `thunderbolt` plugin and `dell-dock` plugin provide same device for TBT3.  In TBT3 implementation, when connected to a USB4/TBT3 port the `thunderbolt` plugin will be used.  When connected to a USB3 port , the `thunderbolt` plugin doesn't get used, but instead a child device gets made from https://github.com/fwupd/fwupd/blob/master/plugins/dell-dock/fu-plugin-dell-dock.c#L106

You raise a good point though in your confusion.  We don't really need https://github.com/fwupd/fwupd/blob/master/plugins/thunderbolt/fu-plugin-thunderbolt.c#L86 anymore.  That was during pre-production that both endpoints were created simultaneously.  But now I2C endpoint is only made when thunderbolt plugin isn't used, so the priority stuff can (and should be) dropped.

Reply:
Added https://github.com/fwupd/fwupd/pull/3333 to clarify that point you raised.